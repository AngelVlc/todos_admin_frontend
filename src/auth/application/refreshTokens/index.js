export * from "./GetRefreshTokensUseCase";
export * from "./DeleteSelectedRefreshTokensUseCase";
