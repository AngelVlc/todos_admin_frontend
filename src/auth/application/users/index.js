export * from "./GetUserByIdUseCase";
export * from "./DeleteUserByIdUseCase";
export * from "./GetUsersUseCase";
export * from "./CreateUserUseCase";
export * from "./UpdateUserUseCase";
