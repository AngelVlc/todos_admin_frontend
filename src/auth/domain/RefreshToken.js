export class RefreshToken {
  constructor({id, userId, expirationDate}) {
    this._id = id;
    this._userId = userId;
    this._expirationDate = expirationDate;
    this._selected = false;
  }

  get id() {
    return this._id;
  }

  get userId() {
    return this._userId;
  }

  get expirationDate() {
    return this._expirationDate;
  }

  get selected() {
    return this._selected;
  }

  set selected(value) {
    this._selected = value;
  } 
}