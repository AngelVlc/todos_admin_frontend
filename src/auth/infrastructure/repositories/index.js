export * from './RefreshTokensRepository';
export * from './UsersRepository';
