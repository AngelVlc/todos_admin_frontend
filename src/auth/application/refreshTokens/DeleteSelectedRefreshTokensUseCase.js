import { BaseUseCase } from "../../../shared/domain/BaseUseCase";
import { RefreshTokensRepository } from "../../infrastructure/repositories";

export class DeleteSelectedRefreshTokensUseCase extends BaseUseCase {
  _repository;

  constructor({ repository }) {
    super();

    this._repository = repository;
  }

  static create() {
    const repository = new RefreshTokensRepository();
    return new DeleteSelectedRefreshTokensUseCase({repository});
  }

  async execute(refreshTokens) {
    const selectedIDs = refreshTokens.reduce((filtered, rt) => {
      if (rt.selected) {
        filtered.push(rt.id);
      }

      return filtered;
    }, []);

    return await this._repository.deleteByIds(selectedIDs);
  }
}
