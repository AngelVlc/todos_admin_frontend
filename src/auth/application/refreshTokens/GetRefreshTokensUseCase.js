import { BaseUseCase } from "../../../shared/domain/BaseUseCase";
import { RefreshToken } from "../../domain";
import { RefreshTokensRepository } from "../../infrastructure/repositories";

export class GetRefreshTokensUseCase extends BaseUseCase {
  _repository;

  constructor({ repository }) {
    super();

    this._repository = repository;
  }

  static create() {
    const repository = new RefreshTokensRepository();

    return new GetRefreshTokensUseCase({ repository });
  }

  async execute() {
    const result = await this._repository.getAll();

    return result.map((item) => {
      return new RefreshToken(item);
    });
  }
}
