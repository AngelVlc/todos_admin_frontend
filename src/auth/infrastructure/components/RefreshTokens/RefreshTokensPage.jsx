import React, { useEffect, useState, useContext, useCallback } from "react";
import { Link } from "react-router-dom";
import { TableWithSelectors } from "../../../../shared/infrastructure/components/TableWithSelectors/TableWithSelectors";
import { AppContext } from "../../../../shared/infrastructure/contexts";
import {
  GetRefreshTokensUseCase,
  DeleteSelectedRefreshTokensUseCase,
} from "../../../application/refreshTokens";

export const RefreshTokensPage = () => {
  const [tokens, setTokens] = useState([]);
  const { useCaseFactory } = useContext(AppContext);
  const [pageInfo, setPageInfo] = useState({
    pageNumber: 1,
    pageSize: 10,
    sortColumn: "id",
    sortOrder: "asc",
  });

  const getRefreshTokens = useCallback(async () => {
    const getRefreshTokensUseCase = useCaseFactory.get(GetRefreshTokensUseCase);
    const refreshTokens = await getRefreshTokensUseCase.execute(pageInfo);

    setTokens(refreshTokens);
  }, [useCaseFactory, pageInfo]);

  useEffect(() => {
    getRefreshTokens();
  }, [getRefreshTokens]);

  const onDeleteSelectedTokens = async () => {
    const deleteSelectedRefreshTokensUseCase = useCaseFactory.get(
      DeleteSelectedRefreshTokensUseCase
    );
    const ok = await deleteSelectedRefreshTokensUseCase.execute(tokens);
    if (ok) {
      getRefreshTokens();
    }
  };

  return (
    <div className="container">
      <h3 className="title">Refresh Tokens</h3>
      <nav className="breadcrumb" aria-label="breadcrumbs">
        <ul>
          <li>
            <a href="/">Home</a>
          </li>
          <li className="is-active">
            <Link aria-current="page" to={`/refreshtokens`}>
              Refresh Tokens
            </Link>
          </li>
        </ul>
      </nav>
      <div className="buttons">
        <button
          className="button is-danger"
          data-testid="deleteSelected"
          onClick={() => onDeleteSelectedTokens()}
        >
          Delete Selected
        </button>
      </div>
      <TableWithSelectors
        columnTitles={["ID", "UserID", "Expiration Date"]}
        columnNames={["id", "userId", "expirationDate"]}
        rows={tokens}
        idColumnName={"id"}
        selectedColumnName={"selected"}
        onSelectedChanged={setTokens}
      />
    </div>
  );
};
