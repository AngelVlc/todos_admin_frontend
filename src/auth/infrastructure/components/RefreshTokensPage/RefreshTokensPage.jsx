import React, { useEffect, useState, useContext, useCallback } from "react";
import { Breadcrumb } from "../../../../shared/infrastructure/components/Breadcrumb";
import { TableWithSelectors } from "../../../../shared/infrastructure/components/TableWithSelectors/TableWithSelectors";
import { AppContext } from "../../../../shared/infrastructure/contexts";
import {
  GetRefreshTokensUseCase,
  DeleteSelectedRefreshTokensUseCase,
} from "../../../application/refreshTokens";

export const RefreshTokensPage = () => {
  const [tokens, setTokens] = useState();
  const { useCaseFactory } = useContext(AppContext);
  const [pageInfo, setPageInfo] = useState({
    pageNumber: 1,
    pageSize: 10,
    sortColumn: "id",
    sortOrder: "desc",
  });

  const getRefreshTokens = useCallback(async () => {
    const getRefreshTokensUseCase = useCaseFactory.get(GetRefreshTokensUseCase);
    const refreshTokens = await getRefreshTokensUseCase.execute(pageInfo);

    setTokens(refreshTokens);
  }, [useCaseFactory, pageInfo]);

  useEffect(() => {
    getRefreshTokens();
  }, [getRefreshTokens]);

  const onDeleteSelectedTokens = async () => {
    const deleteSelectedRefreshTokensUseCase = useCaseFactory.get(
      DeleteSelectedRefreshTokensUseCase
    );
    const ok = await deleteSelectedRefreshTokensUseCase.execute(tokens);
    if (ok) {
      getRefreshTokens();
    }
  };

  return (
    <>
      {tokens && (
        <div className="container">
          <Breadcrumb
            items={[{ url: "/refreshtokens", text: "Refresh Tokens" }]}
          />
          <h3 className="title">Refresh Tokens</h3>
          <div className="buttons">
            <button
              className="button is-danger"
              data-testid="deleteSelected"
              onClick={() => onDeleteSelectedTokens()}
            >
              Delete Selected
            </button>
          </div>
          <TableWithSelectors
            columns={[
              { name: "id", title: "ID" },
              { name: "userId", title: "User ID" },
              { name: "expirationDate", title: "Expiration Date" },
            ]}
            rows={tokens}
            idColumnName={"id"}
            selectedColumnName={"selected"}
            paginationInfo={pageInfo}
            changeSelected={setTokens}
            changePagination={setPageInfo}
          />
        </div>
      )}
    </>
  );
};
