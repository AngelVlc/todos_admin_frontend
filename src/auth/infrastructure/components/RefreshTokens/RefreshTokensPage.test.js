import { render, cleanup, fireEvent, waitFor } from "@testing-library/react";
import { RefreshTokensPage } from "./RefreshTokensPage";
import { AppContext } from "../../../../shared/infrastructure/contexts";
import { createMemoryHistory } from "history";
import { Router } from "react-router-dom";
import { act } from "react-dom/test-utils";
import { GetRefreshTokensUseCase } from "../../../application/refreshTokens";
import { RefreshToken } from "../../../domain";

afterEach(cleanup);

const history = createMemoryHistory();

const refreshTokens = [
  new RefreshToken({
    id: 1,
    userId: 1,
    expirationDate: "2021-01-29T16:46:58Z",
  }),
  new RefreshToken({
    id: 2,
    userId: 2,
    expirationDate: "2021-03-29T16:46:58Z",
  }),
  new RefreshToken({
    id: 3,
    userId: 1,
    expirationDate: "2021-04-29T16:46:58Z",
  }),
];

const mockedGetRefreshTokensUseCase = {
  execute: () => refreshTokens,
};

const mockedDeleteRefreshTokensByIdUseCase = {
  execute: jest.fn(),
};

const useCaseFactory = {
  get: (useCase) => {
    if (useCase == GetRefreshTokensUseCase) {
      return mockedGetRefreshTokensUseCase;
    }

    return mockedDeleteRefreshTokensByIdUseCase;
  },
};

const renderWithContextAndRouter = (component) => {
  const context = { auth: { info: {} }, useCaseFactory };
  return {
    ...render(
      <AppContext.Provider value={context}>
        <Router history={history}>{component}</Router>
      </AppContext.Provider>
    ),
  };
};

it("should match the snapshot", async () => {
  let fragment;
  await act(async () => {
    const { asFragment } = renderWithContextAndRouter(<RefreshTokensPage />);
    fragment = asFragment;
  });
  expect(fragment(<RefreshTokensPage />)).toMatchSnapshot();
});

it("should delete the selected refresh tokens", async () => {
  let container;
  await act(async () => {
    container = renderWithContextAndRouter(<RefreshTokensPage />);
  });

  await waitFor(() => {
    fireEvent.click(container.getByTestId("checkBoxItem1"));
    fireEvent.click(container.getByTestId("checkBoxItem3"));
  });

  mockedDeleteRefreshTokensByIdUseCase.execute.mockResolvedValue(true);

  mockedGetRefreshTokensUseCase.execute = () => [
    new RefreshToken({
      id: 2,
      userId: 2,
      expirationDate: "2021-03-29T16:46:58Z",
    }),
  ];

  const cbItem1 = container.getByTestId("checkBoxItem1");
  const cbItem3 = container.getByTestId("checkBoxItem3");

  await waitFor(() => {
    fireEvent.click(container.getByTestId("deleteSelected"));
  });

  expect(mockedDeleteRefreshTokensByIdUseCase.execute.mock.calls.length).toBe(
    1
  );
  expect(
    mockedDeleteRefreshTokensByIdUseCase.execute.mock.calls[0][0]
  ).toStrictEqual(refreshTokens);
  expect(container.getByTestId("checkBoxItem2")).toBeInTheDocument();

  expect(container.queryByTestId("checkBoxItem1")).not.toBeInTheDocument();
  expect(container.queryByTestId("checkBoxItem3")).not.toBeInTheDocument();
});
