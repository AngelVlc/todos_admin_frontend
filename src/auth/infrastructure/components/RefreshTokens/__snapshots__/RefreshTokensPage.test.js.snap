// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <h3
      class="title"
    >
      Refresh Tokens
    </h3>
    <nav
      aria-label="breadcrumbs"
      class="breadcrumb"
    >
      <ul>
        <li>
          <a
            href="/"
          >
            Home
          </a>
        </li>
        <li
          class="is-active"
        >
          <a
            aria-current="page"
            href="/refreshtokens"
          >
            Refresh Tokens
          </a>
        </li>
      </ul>
    </nav>
    <div
      class="buttons"
    >
      <button
        class="button is-danger"
        data-testid="deleteSelected"
      >
        Delete Selected
      </button>
    </div>
    <div>
      <table
        class="table"
      >
        <thead>
          <tr>
            <td
              data-testid="header-id"
            >
              <span
                class="icon is-inline"
              >
                ID
                <i
                  class="ml-1 fas fa-thin fa-sort-down"
                />
              </span>
            </td>
            <td
              data-testid="header-userId"
            >
              <span
                class="icon is-inline"
              >
                User ID
              </span>
            </td>
            <td
              data-testid="header-expirationDate"
            >
              <span
                class="icon is-inline"
              >
                Expiration Date
              </span>
            </td>
            <td>
              <center>
                <input
                  data-testid="toggleSelectAll"
                  type="checkbox"
                />
              </center>
            </td>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>
              1
            </td>
            <td>
              1
            </td>
            <td>
              2021-01-29T16:46:58Z
            </td>
            <td>
              <center>
                <input
                  data-testid="checkBoxItem1"
                  type="checkbox"
                />
              </center>
            </td>
          </tr>
          <tr>
            <td>
              2
            </td>
            <td>
              2
            </td>
            <td>
              2021-03-29T16:46:58Z
            </td>
            <td>
              <center>
                <input
                  data-testid="checkBoxItem2"
                  type="checkbox"
                />
              </center>
            </td>
          </tr>
          <tr>
            <td>
              3
            </td>
            <td>
              1
            </td>
            <td>
              2021-04-29T16:46:58Z
            </td>
            <td>
              <center>
                <input
                  data-testid="checkBoxItem3"
                  type="checkbox"
                />
              </center>
            </td>
          </tr>
        </tbody>
      </table>
      <div
        class="level"
      >
        <div
          class="level-left"
        >
          <div
            class="ml-3 level-item"
          >
            <input
              class="pageNumber"
              data-testid="pagenumber-input"
              min="1"
              type="number"
              value="1"
            />
            <span
              class="icon"
            >
              <i
                class="fas fa-thin fa-forward"
              />
            </span>
          </div>
          <div
            class="ml-3 level-item"
          >
            <span>
              Per page: 
            </span>
            <select
              data-testid="pagesize-select"
            >
              <option
                value="5"
              >
                5
              </option>
              <option
                value="10"
              >
                10
              </option>
              <option
                value="20"
              >
                20
              </option>
            </select>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
